docs/_ext/validating_code_block.py

```diff
@@ -0,0 +1,46 @@
+from typing import List
+from docutils import nodes
+from docutils.parsers.rst import Directive
+from docutils.parsers.rst import directives
+from sphinx.application import Sphinx
+from sphinx.util.docutils import SphinxDirective
+from sphinx.directives.code import CodeBlock
+from sphinx.errors import ExtensionError
+
+import subprocess
+
+class ValidatingCodeBlock(CodeBlock):
+
+    has_content = True
+    required_arguments = CodeBlock.required_arguments
+    optional_arguments = CodeBlock.optional_arguments
+    final_argument_whitespace = CodeBlock.final_argument_whitespace
+    option_spec = {
+        'type-name': directives.unchanged,
+    }
+    option_spec.update(CodeBlock.option_spec)
+
+    def run(self):
+        source, line = self.state_machine.get_source_and_line(self.lineno)
+        # built-in directives.unchanged_required option validator produces a confusing error message
+        if (self.options.get('type-name') == None):
+            raise ExtensionError("Expected type name in: {0} line: {1}".format(source, line))
+
+        process = subprocess.Popen(['bazel', 'run', '//tools/config_validation:validate_fragment', '--', self.options.get('type-name'), "-s", "\n".join(self.content)],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE)
+        stdout, stderr = process.communicate()
```

Comment:
This looks really nice and clean. You mention on Slack that this is pretty slow (around 10s), which won't scale. You mentioned a "customer builder", curious to learn more. If you want to continue with your existing PR, I'd suggest that you build a `.par` with all dependencies in `docs/build.sh` before invoking Sphinx. Then you can just reference the path of this `.par` here. That will allow you to skip all the Bazel overhead on each YAML processing.

That might still end up being too slow, e.g. if it takes ~1s and you have 100, that's over a minute and a half. When I would do then is add an env var to control validation. In CI we would always enable, but for local iterations on docs builds, we could disable.

Reply:
> You mentioned a "customer builder", curious to learn more

My thinking is that we'd use a dedicated pass to verify example configs by invoking a dedicated builder. All it would do is to validate configs and generate a report. When a "normal" builder is used, examples would still be rendered, but without validation.

Basically we'd be doing config validation in one batch and it would be explicitly invoked. Current implementation combines doc generation and validation, and I'm not sure I can/should continue on validation errors: I think putting a global state into a directive (which is what `ValidatingCodeBlock` is) is a way to go.

Reply:
+1 to failing hard on validation errors.

How about this for an idea.. 
1. We have a Sphinx plugin that just writes out the (YAML fragment, type) tuples to some directory and build all the docs.
2. We then run them all through the `config_validator` in a single `bazel run` at the end?

Reply:
I think what you are suggesting is quite close to a dedicated builder, it would qork quite similar to what you have described. This is the way to go If we want to validate *all* config examples and then report the ones that failed (as opposed to stopping at the first failure, like it's currently implemented).

Reply:
Yeah, I think the other advantage is pure speed. Right now, you are invoking Bazel and Python multiple times to be able to do the validation. I'm guessing only a small fraction of CPU cycles are spent actually in the validation (could be worth measuring).